File: solutions/python/04/bfs01/main.py
from collections import deque


def bfs01(g, start, n):
    INF = 10**18
    dist = [INF] * n
    dist[start] = 0
    dq = deque([start])
    while dq:
        v = dq.popleft()
        for to, cost in g[v]:
            nd = dist[v] + cost
            if nd < dist[to]:
                dist[to] = nd
                if cost == 0:
                    dq.appendleft(to)
                else:
                    dq.append(to)
    return dist


def main() -> None:
    n = 4
    g = [[] for _ in range(n)]
    g[0].append((1, 0))
    g[1].append((0, 0))
    g[1].append((2, 1))
    g[2].append((1, 1))
    g[2].append((3, 0))
    g[3].append((2, 0))
    print(bfs01(g, 0, n))


if __name__ == "__main__":
    main()

File: solutions/go/04/bfs01/main.go
package main

import "fmt"

func bfs01(g [][]struct{ to, cost int }, start, n int) []int {
	const INF = 1 << 60
	dist := make([]int, n)
	for i := range dist {
		dist[i] = INF
	}
	dist[start] = 0
	dq := []int{start}
	for head := 0; head < len(dq); head++ {
		v := dq[head]
		for _, e := range g[v] {
			nd := dist[v] + e.cost
			if nd < dist[e.to] {
				dist[e.to] = nd
				if e.cost == 0 {
					dq = append([]int{e.to}, dq...)
				} else {
					dq = append(dq, e.to)
				}
			}
		}
	}
	return dist
}

func main() {
	n := 4
	g := make([][]struct{ to, cost int }, n)
	add := func(u, v, c int) {
		g[u] = append(g[u], struct{ to, cost int }{v, c})
	}
	add(0, 1, 0)
	add(1, 0, 0)
	add(1, 2, 1)
	add(2, 1, 1)
	add(2, 3, 0)
	add(3, 2, 0)
	fmt.Println(bfs01(g, 0, n))
}
